src/equihash.rs

```diff
@@ -0,0 +1,272 @@
+use blake2b::{Blake2b, Hash, ParameterBlock};
+use byteorder::{LittleEndian, WriteBytesExt};
+
+struct Params {
+    n: u32,
+    k: u32,
+}
+
+#[derive(Clone)]
+struct Node {
+    hash: Vec<u8>,
+    indices: Vec<u32>,
+}
+
+impl Params {
+    fn indices_per_hash_output(&self) -> u32 {
+        512 / self.n
+    }
+    fn hash_output(&self) -> u8 {
+        (self.indices_per_hash_output() * self.n / 8) as u8
+    }
+    fn collision_bit_length(&self) -> usize {
+        (self.n / (self.k + 1)) as usize
+    }
+    fn collision_byte_length(&self) -> usize {
+        (self.collision_bit_length() + 7) / 8
+    }
+    fn hash_length(&self) -> usize {
+        ((self.k as usize) + 1) * self.collision_byte_length()
+    }
+}
+
+impl Node {
+    fn new(a: &mut Node, b: &mut Node, trim: usize) -> Self {
+        let hash: Vec<_> = a.hash
+            .iter()
+            .zip(b.hash.iter())
+            .skip(trim)
+            .map(|(a, b)| a ^ b)
+            .collect();
+        let mut indices = Vec::new();
+        if a.indices_before(b) {
+            indices.append(a.indices.as_mut());
+            indices.append(b.indices.as_mut());
+        } else {
+            indices.append(b.indices.as_mut());
+            indices.append(a.indices.as_mut());
+        }
+        Node {
+            hash: hash,
+            indices: indices,
+        }
+    }
+
+    fn indices_before(&self, other: &Node) -> bool {
+        // Indices are serialized in big-endian so that integer
+        // comparison is equivalent to array comparison
+        self.indices[0] < other.indices[0]
+    }
+
+    fn is_zero(&self, len: usize) -> bool {
+        self.hash.iter().take(len).all(|v| *v == 0)
+    }
+}
+
+fn initialise_state(n: u32, k: u32, digest_len: u8) -> Blake2b {
+    let mut personalization: Vec<u8> = Vec::from("ZcashPoW");
+    personalization.write_u32::<LittleEndian>(n).unwrap();
+    personalization.write_u32::<LittleEndian>(k).unwrap();
+
+    let params = ParameterBlock::new()
+        .set_digest_len(digest_len)
+        .set_key_len(0)
+        .set_fanout(1)
+        .set_max_depth(1)
+        .set_personalization(&personalization);
+    Blake2b::with_parameter_block(digest_len as usize, params)
+}
+
+fn generate_hash(base_state: &Blake2b, i: u32) -> Hash {
+    let mut lei = vec![];
+    lei.write_u32::<LittleEndian>(i).unwrap();
+
+    let mut state = base_state.clone();
+    state.update(&lei);
+    state.finish()
+}
+
+fn has_collision(a: &Node, b: &Node, len: usize) -> bool {
+    a.hash
+        .iter()
+        .zip(b.hash.iter())
+        .take(len)
+        .all(|(a, b)| a == b)
+}
+
+fn distinct_indices(a: &Node, b: &Node) -> bool {
+    for i in &(a.indices) {
```

Comment:
a.indicies.iter().zip(b.indicies.iter()).all(|(a, b)| a != b)

Reply:
AIUI `all()` runs the closure on every item in the iterator, which in this case is pairs of indices `(a.indices[i], b.indices[i])` due to the `zip()`. What we actually want is to run the closure on every possible pair of indices `(a.indices[i], b.indices[j])`.

Reply:
Ah. `itertools::cartesian_product` would be useful here if only I felt good about bringing that into the dependency tree. :)